eos/utils/integrate.cc

```diff
@@ -150,4 +163,110 @@ namespace eos
 
         return result;
     }
+
+    namespace GSL
+    {
+        QNG::Config::Config() :
+            _epsabs(1e-3),
+            _epsrel(1e-4)
+        {
+        }
+
+        double QNG::Config::epsabs() const
+        {
+            return _epsabs;
+        }
+
+        QNG::Config & QNG::Config::epsabs(const double &x)
+        {
+            _epsabs = x;
+            return *this;
+        }
+
+        double QNG::Config::epsrel() const
+        {
+            return _epsrel;
+        }
+
+        QNG::Config & QNG::Config::epsrel(const double &x)
+        {
+            _epsrel = x;
+            return *this;
+        }
+
+        QAGS::Config::Config() :
+            QNG::Config(),
+            _limit(1000),
+            _key(2)
+        {
+        }
+
+        size_t QAGS::Config::limit() const
+        {
+            return _limit;
+        }
+
+        QAGS::Config & QAGS::Config::limit(const size_t & x)
+        {
+            _limit = x;
+            return *this;
+        }
+
+        int QAGS::Config::key() const
+        {
+            return _key;
+        }
+
+        QAGS::Config & QAGS::Config::key(const int & x)
+        {
+            _key = x;
+            return *this;
+        }
+    }
+
+    template <>
+    double integrate<GSL::QNG>(const fdd &f, const double &a, const double &b, const GSL::QNG::Config &config)
+    {
+        double result, abserr;
+        size_t neval;
+        gsl_function F;
+        F.function = &gsl_function_adapter;
+        F.params = (void*)&f;
+
+        auto status = gsl_integration_qng(&F, a, b, config.epsabs(), config.epsrel(),
+                                          &result, &abserr, &neval);
+
+        if (status)
+        {
+            throw IntegrationError(gsl_strerror(status));
+        }
+        return result;
+    }
+
+    template <>
+    double integrate<GSL::QAGS>(const fdd &f, const double &a, const double &b, const GSL::QAGS::Config &config)
+    {
+        double result, abserr;
+        gsl_function F;
+        F.function = &gsl_function_adapter;
+        F.params = (void*)&f;
+
+        // TODO possible memory leak if exception in gsl_integration_qag
+        auto ws = gsl_integration_workspace_alloc(config.limit());
+        auto status = gsl_integration_qag(&F, a, b, config.epsabs(), config.epsrel(),
+                                          config.limit(), config.key(), ws,
+                                          &result, &abserr);
+        gsl_integration_workspace_free(ws);
```

Comment:
Allocating and freeing the workspace everytime this is called seems a bit of a waste of CPU cycles. I was wondering if one can modify ```GSL::QAGS``` as follows:
``` C++
class QAGS
{
    class Workspace
    {
         gsl_integration_workspace * _work_space;
         Workspace() :
             _work_space(gsl_integration_workspace_alloc(...))
         {
             if (! _work_space) throw IntegrationError("Could not allocate ...");
         }
         ~Workspace() { gsl_integration_workspace_free(_work_space); }
         operator=(const Workspace &) = delete;
         operator=(Workspace &&) = delete;
         operator gsl_integration_workspace * () const { return _work_space; }
    };
    static thread_local Workspace work_space;
}
```

You can then use ```QAGS::work_space``` instead of ```ws``` without any explicit allocations.

Reply:
Yup. `static thread_local` is the perfect fit in this case. And RAII of course solves the exception-safety issue.